lale/operators.py

```diff
@@ -3684,6 +3684,65 @@ def _final_individual_op(self) -> Optional["IndividualOp"]:
         else:
             return op._final_individual_op
 
+    def replace(
+        self, original_op: Operator, replacement_op: Operator
+    ) -> "BasePipeline[OpType]":
+        """Replaces an original operator with a replacement operator for the pipeline.
+
+        Parameters
+        ----------
+        original_op :
+            Operator to replace within pipeline.
+
+        replacement_op :
+            Operator to replace the original with.
+
+        Returns
+        -------
+        modified_pipeline :
+            Modified pipeline where original operator is replaced with replacement.
+        """
+
+        new_steps: List[Operator] = []
+        new_edges: List[Tuple[Operator, Operator]] = []
+
+        replacements: Dict[Operator, Operator] = {}
+
+        # NOTE: needs to deal with multiple levels of nesting
+        for step in self.steps():
+            if step == original_op:
```

Comment:
Is there a clean way to tell if a `step` is a trainable or trained version of `original_op` though?

Reply:
As I was looking through the code to answer your question, I noticed that @shinnar recently [overloaded](https://github.com/jsntsay/lale/blob/c5635e2043718ed362054ace58b254cd7b535d18/lale/operators.py#L1911) the `isinstance` check for `IndividualOp`, which might do exactly what we need here.

Reply:
From testing it quickly, I think if I switched to `isinstance` checks, any instance of the operator to replace would be replaced. For example, if you had the following:
```python
pca1 = PCA(n_components=10)
pca2 = PCA(n_components=5)
choice = pca1 | pca2
choice.replace(pca1, SimpleImputer)
```
Then both `pca1` and `pca2` would be replaced. I personally don't mind having this behavior and can just implement and document the function to do this. The alternative would be strict equivalence like I had in the original implementation.

Reply:
I feel like a trainable operator should match the planned operator based on isinstance, since the planned operator is purposely underspecified. On the other hand, two trainable operators with conflicting hyperparameter configurations should not match. Building on your example:
```python
pca1 = PCA(n_components=10)
pca2 = PCA(n_components=5)
choice = pca1 | pca2
choice.replace(pca1, SimpleImputer)  # SimpleImputer | pca2
choice.replace(pca2, SimpleImputer)  # pca1 | SimpleImputer
choice.replace(PCA, SimpleImputer)   # SimpleImputer | SimpleImputer
```